app/grandchallenge/algorithms/views.py

```diff
@@ -1091,3 +1095,61 @@ def get(self, *_, **__):
                 filename=f"{algorithm.slug}-template.zip",
                 content_type="application/zip",
             )
+
+
+class AlgorithmInterfacePermissionMixin(
+    VerificationRequiredMixin, AccessMixin
+):
+    @property
+    def algorithm(self):
+        return get_object_or_404(Algorithm, slug=self.kwargs["slug"])
+
+    def dispatch(self, request, *args, **kwargs):
+        if request.user.has_perm(
+            "change_algorithm", self.algorithm
+        ) and request.user.has_perm("algorithms.add_algorithm"):
+            return super().dispatch(request, *args, **kwargs)
+        else:
+            return self.handle_no_permission()
+
+
+class AlgorithmInterfaceForAlgorithmCreate(
+    AlgorithmInterfacePermissionMixin, CreateView
+):
+    model = AlgorithmInterface
+    form_class = AlgorithmInterfaceGetOrCreateForm
+    success_message = "Algorithm interface successfully added"
```

Comment:
I was unsure which `model` to use for this view. This view either creates or retrieves an existing `AlgorithmInterface` and then adds it to an algorithm, i.e. adds a row to the custom through table `AlgorithmAlgorithmInterface`. So at the end of the day, it creates an entry in the through model, but it feels wrong to have a `CreateView` for a through model, since that is not normally accessible. So I ultimately went with a `CreateView` for the `AlgorithmInterface`, with the name reflecting the side effect that it also adds an entry to the through table. 

Using the `AlgorithmInterface` as base model for the view also interfaces better with the form I'm using. For the form, I have a clear preference for using a `AlgorithmInterface` `ModelForm` (rather than a `ModelForm` for the through model).  Much more straight-forward to implement. 

I also briefly considered making this an `UpdateView` for the `Algorithm` (mainly for the permission check), but I abandoned that pretty quickly.

Reply:
My gut feeling is that `AlgorithmInterface` Create is correct here, or maybe a `FormView` with a custom save method on the form would be more accurate, but I don't think that really matters. Although we have access to the through table as we are using a custom one Django always prefers you not to interact with it directly.

Reply:
I opted to stick to an `AlgorithmInterface` Create.